lib/protos.js

```diff
@@ -109,30 +125,60 @@ exports.track = function(track){};
  * // => []
  */
 
-exports.map = function(events, event){
-  var normalizedEvent = normalize(event);
-
-  return foldl(function(matchingEvents, val, key, events) {
-    // If true, this is a `mixed` value, which is structured like so:
-    //     { key: 'testEvent', value: { event: 'testEvent', someValue: 'xyz' } }
-    // We need to extract the key, which we use to match against
-    // `normalizedEvent`, and return `value` as part of `matchingEvents` if that
-    // match succeds.
-    if (type(events) === 'array') {
-      // If there's no key attached to this event mapping (unusual), skip this
-      // item.
-      if (!val.key) return matchingEvents;
-      // Extract the key and value from the `mixed` object.
-      key = val.key;
-      val = val.value;
+exports.map = function(mapping, key){
+  var normalizedComparator = normalize(key);
+  var mappingType = getMappingType(mapping);
+
+  function getMappingType(mapping) {
+    if (type(mapping) === 'object' && every(isString, mapping)) return 'map';
+    if (type(mapping) === 'array' && every(isString, mapping)) return 'array';
+    if (type(mapping) === 'array' && every(isMixed, mapping)) return 'mixed';
+    return 'unknown';
+  }
+
+  function isString(item) {
+    return type(item) === 'string';
+  }
+
+  /* eslint-disable no-extra-parens */
```

Comment:
sounds good. in fairness: multi-line returns are pretty hot nowadays with jsx and functional react component stuff. thought this was **much** more legible than one super long line